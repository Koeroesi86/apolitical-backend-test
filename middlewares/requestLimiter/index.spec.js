const rateLimiterMiddleware = require('./index');
const { createRequest, createResponse } = require("node-mocks-http");

const provider = {
  get: jest.fn(() => Promise.resolve(0)),
  set: jest.fn(() => Promise.resolve()),
};

describe('articleProvider middleware', () => {
  const middleware = rateLimiterMiddleware(provider);

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should call next by default', async () => {
    const request = createRequest({
      ip: '127.0.0.1',
    });
    const response = createResponse();
    const next = jest.fn();

    await middleware(request, response, next);

    expect(next).toHaveBeenCalledTimes(1);
    expect(provider.set.mock.calls).toMatchSnapshot('provider.set');
    expect(response._getHeaders()).toMatchSnapshot('response headers');
  });

  it('should not allow 4th request', async () => {
    provider.get.mockResolvedValueOnce(3);
    const request = createRequest({
      ip: '127.0.0.1',
    });
    const response = createResponse();
    const next = jest.fn();

    await middleware(request, response, next);

    expect(next).toHaveBeenCalledTimes(0);
    expect(provider.set.mock.calls).toMatchSnapshot('provider.set');
    expect(response._getHeaders()).toMatchSnapshot('response headers');
  });
})
