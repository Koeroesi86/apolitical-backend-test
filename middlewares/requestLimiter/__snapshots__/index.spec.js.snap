// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`articleProvider middleware should call next by default: provider.set 1`] = `
Array [
  Array [
    "127.0.0.1",
    1,
  ],
]
`;

exports[`articleProvider middleware should call next by default: response headers 1`] = `
Object {
  "x-current-count": 1,
}
`;

exports[`articleProvider middleware should not allow 4th request: provider.set 1`] = `
Array [
  Array [
    "127.0.0.1",
    4,
  ],
]
`;

exports[`articleProvider middleware should not allow 4th request: response headers 1`] = `
Object {
  "x-current-count": 4,
}
`;
